### PR TITLE
[Inductor] MixOrderReduction support low precision reduction

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #169978

The issue does not appear for RMSNorm/LayerNorm since even if input is in low precision, RMSNorm will cast the inputs to FP32 first and do a down cast in the end. The reduction output is actually in FP32.

This PR fixes the issue if the reduction output is in low precision. Better to fix it to make MixOrderReduction a general feature.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/codegen/simd.py [hunk 1]
```diff
@@ -1748,9 +1748,12 @@ def _bench(candidate_split_size):
                 partial_accum.reduction_type, partial_accum.reduction_type
             )
 
-            V.graph.wrapper_code.writeline(
-                f"{buffer_name} = {ws_name}[{start} : {end}].view({nsplit}, {rnumel}).{opname}(dim=0)",
-            )
+            final_reduce = f"{buffer_name} = {ws_name}[{start} : {end}].view({nsplit}, {rnumel}).{opname}(dim=0)"
```
Comment (eellison) [inline torch/_inductor/codegen/simd.py:1751]: can we update the final_reduce so that we allocate it in the right number of bytes to begin with ? instead of a clone below

File: torch/_inductor/codegen/simd.py [hunk 2]
```diff
@@ -1748,9 +1748,12 @@ def _bench(candidate_split_size):
...
+            # The workspace tensor is in torch.float, need a cast if the buffer is
+            # not.
+            if (buffer_dtype := V.graph.get_dtype(buffer_name)) != torch.float:
```
Comment (eellison) [inline torch/_inductor/codegen/simd.py:1754]: I don't think it's strictly `float`. its more that its low precision and accum is done in fp32. i did a pr for this for split reductions, let me try to find

Comment (eellison) [reply torch/_inductor/codegen/simd.py:1754]: https://github.com/pytorch/pytorch/blob/171ae4a8d20781d2853126f6913accfa291f7e30/torch/_inductor/ir.py#L1897-L1903

but i think its just low precision dtype float dtype -> fp32

File: torch/_inductor/codegen/simd.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/codegen/simd.py:1751]: We still want to do partial reduction in high precision.

File: torch/_inductor/codegen/simd.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/codegen/simd.py:1754]: We allocate the workspace tensor in fp32. So if the buffer has a different dtype, we should cast.

Is your concern about even higher precision like fp64 inputs?

File: torch/_inductor/codegen/simd.py (hunk 1 above)
Comment (eellison) [reply torch/_inductor/codegen/simd.py:1751]: Can we allocate temporary in fp32, write to it in fp32, then sum -> output in lower precision ?

Comment (shunting314) [reply torch/_inductor/codegen/simd.py:1751]: Like calling `torch.sum(x, dtype=torch.bfloat16)` directly? Instead of `torch.sum(x).to(dtype=torch.bfloat16)`

Comment (eellison) [reply torch/_inductor/codegen/simd.py:1751]: or using an inductor kernel.

Comment (shunting314) [reply torch/_inductor/codegen/simd.py:1751]: I'm not sure if eager would truncate the input first if we do `torch.sum(x, dtype=torch.bfloat16)` . On the other hand, these tensors are all very small since they are partially reduced tensors.

Comment (eellison) [reply torch/_inductor/codegen/simd.py:1751]: yea, if it's small enough where the memory savings dont matter, deferring to you

Comment (shunting314) [reply torch/_inductor/codegen/simd.py:1751]: I'm a bit hesitated to use a Inductor kernel here since I don't like any of the following alternatives
1. call torch.compile: should we ever recursively do torch.compile?
2. do lowering and generate new buffers: are we supposed to do lowering in scheduler. This seems mess up code layers.

For now, probably just calling into eager is the cleanest choice. The buffers are small anyways

File: torch/_inductor/codegen/simd.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/codegen/simd.py:1754]: yes, fp64 shouldnt downcast to fp32

Comment (shunting314) [reply torch/_inductor/codegen/simd.py:1754]: I see. For fp64, we should allocate the workspace buffer in fp64 in the first place

Comment (eellison) [reply torch/_inductor/codegen/simd.py:1754]: but for f16 as well, we shouldnt write to it in fp32, or it will truncate on store.

**Review by eellison — APPROVED** [review]
